lib/driver.js

```diff
@@ -43,29 +43,39 @@ class AndroidDriver extends BaseDriver {
   }
 
   async createSession (caps) {
-    let sessionId;
-    [sessionId] = await super.createSession(caps);
-
-    // TODO fail if app file is missing, or if app is not on the device?
-    // assigning defaults
-    let defaultOpts = {action: "android.intent.action.MAIN",
-                       category: "android.intent.category.LAUNCHER",
-                       flags: "0x10200000",
-                       disableAndroidWatchers: false,
-                       tmpDir: await tempDir.staticDir(),
-                       fullReset: false};
-    _.defaults(this.opts, defaultOpts);
-    if (!this.opts.javaVersion) {
-      this.opts.javaVersion = await helpers.getJavaVersion();
-    }
-    // Add condition for chrome driver
-    if (this.opts.app) {
-      // download and unzip
-      this.opts.app = await this.helpers.configureApp(this.opts.app, APP_EXTENSION);
+    // the whole createSession flow is surrounded in a try-catch statement
+    // if creating a session fails at any point, we teardown everything we
+    // set up before throwing the error.
+    try {
+
+      let sessionId;
+      [sessionId] = await super.createSession(caps);
+
+      // TODO fail if app file is missing, or if app is not on the device?
+      // assigning defaults
+      let defaultOpts = {action: "android.intent.action.MAIN",
+                         category: "android.intent.category.LAUNCHER",
+                         flags: "0x10200000",
+                         disableAndroidWatchers: false,
+                         tmpDir: await tempDir.staticDir(),
+                         fullReset: false};
+      _.defaults(this.opts, defaultOpts);
+      if (!this.opts.javaVersion) {
+        this.opts.javaVersion = await helpers.getJavaVersion();
+      }
+      // Add condition for chrome driver
```

Comment:
fyi, this is going to conflict with my PR as well, guess we'll get to see whose is merged first :-)